### DISCUSSION TITLE
Are Discard and Return not allowed in the same function as a textureSample?

### DISCUSSION THREAD

**Discussion by jonathandw743** [gfx-rs/wgpu#3695] [Q&A]
The following code produces the error:

```bash
    Function [4] 'col_from_hit_pos' is invalid
    Required uniformity of control flow for IMPLICIT_LEVEL in [19] is not fulfilled because of Return
```

```wgsl
fn col_from_hit_pos(hit_pos: vec3<f32>) -> vec3<f32> {
    if abs(hit_pos.x) > 0.5 || abs(hit_pos.z) > 0.5 {
        discard;
    }
    if abs(hit_pos.x) > 0.49 || abs(hit_pos.z) > 0.49 {
        return vec3<f32>(1.0);
    }
    return textureSample(t_diffuse, s_diffuse, hit_pos.xz + 0.5).xyz;
}
```

But when I wrap the textureSample in its own function (as in the following code), the error stops showing and the shader works as intended.

```wgsl
fn texture_sample_at_hit_pos(hit_pos: vec3<f32>) -> vec3<f32> {
    return textureSample(t_diffuse, s_diffuse, hit_pos.xz + 0.5).xyz;
}

fn col_from_hit_pos(hit_pos: vec3<f32>) -> vec3<f32> {
    if abs(hit_pos.x) > 0.5 || abs(hit_pos.z) > 0.5 {
        discard;
    }
    if abs(hit_pos.x) > 0.49 || abs(hit_pos.z) > 0.49 {
        return vec3<f32>(1.0);
    }
    return texture_sample_at_hit_pos(hit_pos);
}
```

Is this intended behaviour?
And if so may I ask why?

**Comment by victorvde** [comment]
Rejecting the first example is correct, [textureSample](https://www.w3.org/TR/WGSL/#texturesample) needs to be called from uniform control flow. The gradient from adjacent pixels is needed for the mip-level and filtering.
The easy solution is to do the textureSample before the if statements.

Putting the textureSample in a function should still be an error, but I think naga doesn't fully implement uniformity analysis yet: https://github.com/gfx-rs/naga/issues/1744

**Reply by jonathandw743** [reply]
So does anything bad actually happen from doing it this way? If not then why should naga flag this?
